File: LC0236_Lowest_Common_Ancestor_Binary_Tree/lca.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func rLCA(root, p, q *TreeNode) *TreeNode {
	if root == nil || root == p || root == q {
		return root
	}
	left := rLCA(root.Left, p, q)
	right := rLCA(root.Right, p, q)
	if left != nil && right != nil {
		return root
	}

	if left != nil {
		return left
	}
	return right
}

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	return rLCA(root, p, q)
}

File: LC0236_Lowest_Common_Ancestor_Binary_Tree/lca.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def rLCA(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if root == None or p == root or q == root:
            return root
        # Determine whether p or q exist in the subtree rooted at root.
        left = self.rLCA(root.left, p, q)
        right = self.rLCA(root.right, p, q)
        if left and right:
            # If we find both p and q in the subtree rooted at root, then root is the answer.
            return root
            
        return left or right

    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        return self.rLCA(root, p, q)